File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
class Solution {
    public int solution(String dartResult) {
    int answer = 0;
      
      char[] dart = dartResult.toCharArray();
      int[] result = new int[3];
      
      int sum = 0; // 각 다트 점수 합
      int dartCnt = 0; // 다트 수 1 2 3
      
      for(int i=0; i<dartResult.length(); i++) {
          if(Character.isDigit(dart[i])) { // 숫자일 경우
              if(i!=0) {
                  result[dartCnt++] = sum;
              }
              if((int)dart[i+1] == '0') { // 10일 경우
                  sum = 10;
                  i++; // 다음 단계 pass
                  continue;
              }
              else
                  sum = dart[i] - '0';
          }
          
          if(dart[i] == 'S') {
              ;
          }
          else if(dart[i] == 'D') {
              sum *= sum;
          }
          else if(dart[i] == 'T') {
              sum = (int)Math.pow(sum, 3); // Math.pow의 리턴형은 double
          }
          else if(dart[i] == '*') {
              sum *= 2;
              if(dartCnt != 0) { // 바로 앞에꺼도 2배
                  result[dartCnt-1] *= 2;
              }
          }
          else if(dart[i] == '#') {
              sum *= -1;
          }
          
          if(i==dartResult.length()-1 && dartCnt==2) // 마지막
              result[dartCnt] = sum;
      }
        
      // 다 더함
      for(int i : result) {
          answer += i;
      }
      return answer;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    
    const check = /[0-9]/;
    let sum = 0;
    let result = [];
    let count = -1;
    
    for(let i = 0; i<dartResult.length; i++){
        
        if(check.test(dartResult[i])){
            
            if(dartResult[i+1] === '0'){
                sum += 10;
                i++
            }else{
                sum += (dartResult[i]);
            }
            result.push(sum);
            sum = 0;
            count++;
            continue;
        }
        
        if(dartResult[i] === 'D'){ //2제곱
            result[count] = Math.pow(result[count], 2);
        }else if(dartResult[i] === 'T'){ //3제곱
            result[count] = Math.pow(result[count], 3);
        }else if(dartResult[i] === '*'){ //스타상 --> 해당 점수와 바로 전에 얻은 점수 2배
            if(count === 0){
                result[count] *= 2;
            }else{
                result[count-1] *= 2;
                result[count] *= 2;
            }
        }else if(dartResult[i] === '#'){ //아차상 --> 해당 점수 -1배, 스타상과 중첩 가능
            result[count] *= -1;
        }
        
        
    }
    
    let answer = 0;
    
    for(let i of result){
        answer += parseInt(i);
    }
    
    return answer;
}